src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+            var httpUrlHostComponent = httpRequestUri.Host;
+            if (!httpRequestUri.IsDefaultPort)
+                httpUrlHostComponent = $"{httpUrlHostComponent}:{httpRequestUri.Port}";
+
+            payload.Add(SignedHttpRequestClaimTypes.U, httpUrlHostComponent);
+        }
+
+        /// <summary>
+        /// Adds the 'm' claim to the <paramref name="payload"/>.
+        /// </summary>
+        /// <param name="payload">HttpRequest payload represented as a <see cref="Dictionary{TKey, TValue}"/>.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <remarks>
+        /// This method will be executed only if <see cref="SignedHttpRequestCreationParameters.CreateP"/> is set to <c>true</c>.
+        /// </remarks>  
+        protected virtual void AddPClaim(Dictionary<string, object> payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var httpRequestUri = signedHttpRequestDescriptor.HttpRequestData.Uri;
+
+            if (httpRequestUri == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor.HttpRequestData.Uri));
+
+            httpRequestUri = EnsureAbsoluteUri(httpRequestUri);
+
+            payload.Add(SignedHttpRequestClaimTypes.P, httpRequestUri.AbsolutePath);
+        }
+
+        /// <summary>
+        /// Adds the 'q' claim to the <paramref name="payload"/>.
+        /// </summary>
+        /// <param name="payload">HttpRequest payload represented as a <see cref="Dictionary{TKey, TValue}"/>.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <remarks>
+        /// This method will be executed only if <see cref="SignedHttpRequestCreationParameters.CreateQ"/> is set to <c>true</c>.
+        /// </remarks>  
+        protected virtual void AddQClaim(Dictionary<string, object> payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var httpRequestUri = signedHttpRequestDescriptor.HttpRequestData.Uri;
+
+            if (httpRequestUri == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor.HttpRequestData.Uri));
+
+            httpRequestUri = EnsureAbsoluteUri(httpRequestUri);
+            var sanitizedQueryParams = SanitizeQueryParams(httpRequestUri);
+
+            StringBuilder stringBuffer = new StringBuilder();
+            List<string> queryParamNameList = new List<string>();
+            try
+            {
+                var lastQueryParam = sanitizedQueryParams.LastOrDefault();
```

Comment:
Perf:
wouldn't it be faster to instead have a first query parameter bool switch ?

Dirty example:

```               
                var firstQueryParam = true;
 
                foreach (var queryParam in sanitizedQueryParams)
                {
		       if (!firstQueryParam)
		       {
			  stringBuffer.Append("&");
		       }
					
                    queryParamNameList.Add(queryParam.Key);
                    var encodedValue = $"{queryParam.Key}={queryParam.Value}";

                    stringBuffer.Append(encodedValue);
					
					firstQueryParam = false;
                }
```

Reply:
I agree, it's simpler too https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1260/commits/8898f83d7adde7173222729b31c1338f164baf01